features/preferences/impl/src/main/kotlin/io/element/android/features/preferences/impl/labs/LabsPresenter.kt

```diff
@@ -0,0 +1,124 @@
+    private val featureFlagService: FeatureFlagService,
+    private val clearCacheUseCase: ClearCacheUseCase,
+) : Presenter<LabsState> {
+    @Composable
+    override fun present(): LabsState {
+        val coroutineScope = rememberCoroutineScope()
+        val features = remember {
+            val entries = featureFlagService.getAvailableFeatures()
+                .filter { it.isInLabs && !it.isFinished }
+                .map { it.key to it }
+            mutableStateMapOf(*entries.toTypedArray())
+        }
+        val enabledFeatures = remember {
+            mutableStateMapOf<String, Boolean>()
+        }
+
+        LaunchedEffect(Unit) {
+            for (feature in features.values) {
+                val isEnabled = featureFlagService.isFeatureEnabled(feature)
+                enabledFeatures[feature.key] = isEnabled
+            }
+        }
+
+        var isApplyingChanges by remember { mutableStateOf(false) }
+
+        val featureUiModels = createUiModels(features, enabledFeatures)
+
+        fun handleEvent(event: LabsEvents) {
+            when (event) {
+                is LabsEvents.ToggleFeature -> coroutineScope.launch {
+                    val feature = features[event.feature] ?: return@launch
+                    val isEnabled = featureFlagService.isFeatureEnabled(feature)
+                    featureFlagService.setFeatureEnabled(feature = feature, enabled = !isEnabled)
+                    enabledFeatures[feature.key] = !isEnabled
+
+                    when (feature.key) {
+                        FeatureFlags.Threads.key -> {
+                            // Threads require a cache clear to recreate the event cache
+                            clearCacheUseCase()
+                            isApplyingChanges = true
+                        }
+                    }
+                }
+            }
+        }
+
+        return LabsState(
+            features = featureUiModels,
+            isApplyingChanges = isApplyingChanges,
+            eventSink = ::handleEvent,
+        )
+    }
+
+    @Composable
+    private fun createUiModels(
+        features: SnapshotStateMap<String, Feature>,
+        enabledFeatures: SnapshotStateMap<String, Boolean>
+    ): ImmutableList<FeatureUiModel> {
+        return features.values.map { feature ->
+            key(feature.key) {
+                val isEnabled = enabledFeatures[feature.key].orFalse()
+                val title = when (feature) {
+                    FeatureFlags.Threads -> stringProvider.getString(R.string.screen_labs_enable_threads)
+                    else -> feature.title
+                }
+                val description = when (feature) {
+                    FeatureFlags.Threads -> stringProvider.getString(R.string.screen_labs_enable_threads_description)
+                    else -> feature.description
+                }
+                val icon = when (feature) {
+                    FeatureFlags.Threads -> CompoundIcons.Threads()
```

Comment:
I would have let the View decide which string resource / Icon to use depending on the feature type, but this code works too, even if it looks a bit harder to read (to me!)